doc/developer/design/20240411_graceful_reconfiguration_of_managed_clusters_v1.md

```diff
@@ -0,0 +1,364 @@
+performs waits/validations.
+
+The new `sequence_alter_cluster` function will need to perform
+the current validations, used in both `sequence_alter_cluster` and
+`sequence_alter_cluster_managed_to_managed`, but will kick off a new
+`StageResult::Handle` when an `alter_mechanism` is provided. The task in this
+handle will perform waits/validations and pass off the finalization to
+a `Finalize` stage.
+
+Additionally, we'll need to perform some extra validations to ensure the
+the `alter_mechanism` and `AlterClusterPlan` are compatible. These are:
+1. validate the cluster is managed
+2. validate that the cluster is not undergoing a reconfiguration
+3. validate the cluster has no sources and now sinks
+4. the cluster cannot have a refresh schedule (v1)
+
+The `sequence_alter_cluster_managed_to_managed` method will need
+to be updated to apply the correct name/pending status for new replicas
+and only remove existing replicas when no alter_mechanism is provided.
+
+
+For V2:
+We will need to implement a message strategy that kicks the work off
+off of the `sequence_staged` flow. In this scenario, we will still create new
+replicas in the foreground, but the wait/checks will be performed in a separate
+message handle by emitting a `ClusterReconfigureCheckReady` message.
+
+The state for the reconfiguration will largely be stored in a `PendingClusterReconfiguration`
+struct.
+
+
+```rust
+struct PendingClusterReconfiguration {
+ ctx: ExecuteContext,
+ // needed to determine timeouts
+ start_time: Instant,
+ // the plan will hold the cluster_id, config, mechanism, etc...
+ plan: AlterClusterPlan,
+ // idempotency key will need to be part of the  message and will need to be
+ // validated to ensure we're looking at the correct plan, in case we
+ // miss a cancel and a new entry was put in pending_cluster_reconfiguration.
+ idempotency_key: Uuid,
+}
+```
+
+In order to cancel we'll keep a map, `pending_cluster_reconfigurations`,  of
+session `ClusterIds` to `PendingClusterReconfigurations`.
+
+*Currently due to limitations in sources and sinks, we will need to maintain
+the disruptive alter mechanism. Once multiple replicas are allowed for sources
+and sinks we can remove the disruptive mechanism and default the wait to `WAIT=For 0 seconds`.
+
+
+#### Handling PendingClusterReconfigure Messages
+
+The responsibilities of this operation will be:
+1. checking for cancellations
+2. checking for the `WAIT` condition
+3. finalizaing the reconfiguration
+4. returning to the user
+
+__Checking the `WAIT` Condition__
+The message handler will perform all checks on every run. If the wait condition
+has not yet been met, we will reschedule the message to be resent. For each
+reschedule we'll spawn a task that waits some duration and sends a
+`ClusterReconfigureCheckReady` message with the connection_id.
+For `WAIT FOR` we can wait for the exact duration that would cause the next check to succeed.
+For `WAIT UNTIL CAUGHT UP` we can just wait for three seconds.
```

Comment:
Add an open question about if there's a better way to do this, for example starting a subscribe and waiting off-thread for that to reach some condition.